OperationBasedWithInterfaces (#4370)
apollographql/apollo-kotlin: design-docs/operationBased2.md

</pre></td>
<td><pre lang="kotlin">
public data class Data(public val animal: Animal?)
data class Animal(
  val lionFragment: LionFragment?,
  val catFragment: CatFragment?,
)
</pre></td>
</tr>
</table>

## Example 3

<table>
<tr><th>GraphQL</th><th>Kotlin</th></tr>
<tr>
<td><pre lang="graphql">
query TestOperation {
  animal {
    # Field type = Inline fragment type
    ... on Animal {
      species
    }
  }
}
</pre></td>
<td><pre lang="kotlin">
public data class Data(public val animal: Animal?)
data class Animal(val onAnimal: OnAnimal)
</pre></td>
</tr>
</table>

## Example 4

<table>
<tr><th>GraphQL</th><th>Kotlin</th></tr>
<tr>
<td><pre lang="graphql">
query TestOperation {
  cat {
    # Field type Cat is subtype of inline fragment
    ... on Animal {
      species
    }
  }
}
</pre></td>
<td><pre lang="kotlin">
public data class Data(public val cat: Cat?)
data class Cat(val onAnimal: OnAnimal)
</pre></td>
</tr>
</table>

## Example 5

<table>
<tr><th>GraphQL</th><th>Kotlin</th></tr>
<tr>
<td><pre lang="graphql">
query TestOperation {
  animal {
    ... on Cat {
      species
    }
  }
}
</pre></td>
<td><pre lang="kotlin">
public data class Data(public val animal: IAnimal?)

> sealed interface IAnimal { val onCat: Animal.OnCat? }
> data class CatAnimal(val onCat: Animal.OnCat): IAnimal
> data class Animal(val onCat: Animal.OnCat?): IAnimal

Reviewer: For consistency with responseBased codegen:

```suggestion
sealed interface Animal { val onCat: Animal.OnCat? }
data class CatAnimal(val onCat: Animal.OnCat): Animal
data class OtherAnimal(val onCat: Animal.OnCat?): Animal
```

Author: OK. 

At first, I thought that it would be more compatible with `operationBased` if I named classes like these. But I was wrong. Your suggestion is more compatible with `operationBased` codegen.

</pre></td>
</tr>
</table>

## Example 6

<table>
<tr><th>GraphQL</th><th>Kotlin</th></tr>
<tr>
<td><pre lang="graphql">
query TestOperation {
  animal {
    species
    ... on WarmBlooded {
      temperature
    }
    ... on Pet {
      name
    }
  }
}
</pre></td>
<td><pre lang="kotlin">
public data class Data(public val animal: IAnimal?)

> // Abstraction Model

Reviewer: ```suggestion
// Interface Model
```

> sealed interface IAnimal { val onAnimal: Animal.IOnAnimal }
> // 3 TypeSet Models
> data class WarmBloodedAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
> data class PetAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
> data class WarmBloodedPetAnimal(val onAnimal: Animal.IOnAnimal): IAnimal
> // Fallback Model
> data class Animal(val onAnimal: IOnAnimal): IAnimal

Reviewer: ```suggestion
sealed interface Animal { 
  val onWarmBlooded: Animal.OnWarmBlooded?
  val onPet: Animal.OnPet?
}
// 3 TypeSet Models
data class WarmBloodedAnimal(
  val onWarmBlooded: Animal.OnWarmBlooded
  val onPet: Animal.OnPet?  
): Animal
data class PetAnimal(
  val onWarmBlooded: Animal.OnWarmBlooded?
  val onPet: Animal.OnPet
): Animal
data class WarmBloodedPetAnimal(
  val onWarmBlooded: Animal.OnWarmBlooded
  val onPet: Animal.OnPet
): Animal
// Fallback Model
data class OtherAnimal(
  val onWarmBlooded: Animal.OnWarmBlooded?
  val onPet: Animal.OnPet?  
): Animal
```

</pre></td>
</tr>
</table>

## Example 7

<table>
<tr><th>GraphQL</th><th>Kotlin</th></tr>
<tr>
<td><pre lang="graphql">
query TestOperation {
  animal {
    ... on Animal {
      species
      # Nested WarmBlooded
      ... on WarmBlooded {
        temperature
      }
      # Nested Pet
      ... on Pet {
        name
      }
    }
  }
}
</pre></td>
<td><pre lang="kotlin">

> // animal is not generated as polymorphic model because inline fragment ... on Animal is always assignable

Reviewer: ```suggestion
// animal is not generated as polymorphic model because the `... on Animal` inline fragment type condition is always satisfied
```

public data class Data(public val animal: Animal?)
data class Animal(val onAnimal: IOnAnimal)

// onAnimal field has two nested inline fragments -> generated as Polymorphic model
public sealed interface IOnAnimal { 
  val species: String,

>   val onWarmBlooded: OnAnimal.OnWarmBlooded?,

Reviewer: I think there's no need to namespace `OnWarmBlooded` here as it will be accessible in the `OnAnimal` scope. `val onWarmBlooded: OnWarmBlooded?` should work (I included it in the suggestion below)

> public data class Data(public val animal: Animal?)
> data class Animal(val onAnimal: IOnAnimal)
> 
> // onAnimal field has two nested inline fragments -> generated as Polymorphic model
> public sealed interface IOnAnimal { 
>   val species: String,
>   val onWarmBlooded: OnAnimal.OnWarmBlooded?,
>    val onPet: OnAnimal.OnPet?
> }
> data class WarmBloodedOnAnimal(
>   val species: String, 
>   val onWarmBlooded: OnAnimal.OnWarmBlooded,
>   val onPet: OnAnimal.OnPet?
> ): IOnAnimal
> data class PetOnAnimal(
>   val species: String,
>   val onWarmBlooded: OnAnimal.OnWarmBlooded?,
>   val onPet: OnAnimal.OnPet,
> ): IOnAnimal
> data class WarmBloodedPetOnAnimal(
>   val species: String,
>   val onWarmBlooded: OnAnimal.OnWarmBlooded,
>   val onPet: OnAnimal.OnPet,
> ): IOnAnimal
> data class OnAnimal(
>   val species: String,
>   val onWarmBlooded: OnWarmBlooded?,
>   val onPet: OnPet?,
> ): IOnAnimal

Reviewer: ```suggestion
public data class Data(public val animal: Animal?)
data class Animal(val onAnimal: OnAnimal)

public sealed interface OnAnimal { 
  val species: String,
  val onWarmBlooded: OnAnimal.OnWarmBlooded?,
  val onPet: OnAnimal.OnPet?
}
data class WarmBloodedOnAnimal(
  val species: String, 
  val onWarmBlooded: OnAnimal.OnWarmBlooded,
  val onPet: OnAnimal.OnPet?
): OnAnimal
data class PetOnAnimal(
  val species: String,
  val onWarmBlooded: OnAnimal.OnWarmBlooded?,
  val onPet: OnAnimal.OnPet,
): OnAnimal
data class WarmBloodedPetOnAnimal(
  val species: String,
  val onWarmBlooded: OnAnimal.OnWarmBlooded,
  val onPet: OnAnimal.OnPet,
): OnAnimal
data class OtherOnAnimal(
  val species: String,
  val onWarmBlooded: OnWarmBlooded?,
  val onPet: OnPet?,
): OnAnimal
```

</pre></td>
</tr>
</table>

## Example 8

<table>
<tr><th>GraphQL</th><th>Kotlin</th></tr>
<tr>
<td><pre lang="graphql">
query TestOperation {
  animal {
    ... AnimalFragment
  }
}
fragment AnimalFragment on Animal {
  species
  ... on WarmBlooded {
    temperature
  }
  ... on Pet {
    name
  }
}
</pre></td>
<td><pre lang="kotlin">

> // animal is not generated as polymorphic model because there is no inline fragment

Reviewer: Can you elaborate why you only want this new codegen for inline fragments and not fragment spreads? With something like this:

```graphql
{
  animal {
    ...catFragment
    ...lionFragment
  }
}
```

it feels like the benefits would be the same as with 

```graphql
{
  animal {
    ... on Cat { ... }
    ... on Lion { ... }
  }
}
```

Any reason to only do the transformation for inline fragments?

> // animal is not generated as polymorphic model because there is no inline fragment
> data class Animal(val animalFragment: IAnimalFragment)
> 
> // AnimalFragment is generated as polymorphic model because it contains inline fragments
> sealed interface IAnimalFragment { 
>   val species: String
>   val onWarmBlooded: AnimalFragment.OnWarmBlooded?
>   val onPet: AnimalFragment.OnPet?
> }
> data class WarmBloodedAnimalFragment(
>   val species: String,
>   val onWarmBlooded: AnimalFragment.OnWarmBlooded,
>   val onPet: AnimalFragment.OnPet?
> ): IAnimalFragment
> data class PetAnimalFragment(
>   val species: String,
>   val onWarmBlooded: AnimalFragment.OnWarmBlooded?,
>   val onPet: AnimalFragment.OnPet
> ): IAnimalFragment
> data class WarmBloodedPetAnimalFragment(
>   val species: String,
>   val onWarmBlooded: AnimalFragment.OnWarmBlooded,
>   val onPet: AnimalFragment.OnPet
> ): IAnimalFragment
> data class AnimalFragment(
>   val species: String,
>   val onWarmBlooded: AnimalFragment.OnWarmBlooded?,
>   val onPet: AnimalFragment.OnPet?
> ): IAnimalFragment

Reviewer: This is basically the same as example 7 with `OnAnimal` replaced with `AnimalFragment`?

```suggestion
public data class Data(public val animal: Animal?)
data class Animal(val animalFragment: AnimalFragment)

public sealed interface AnimalFragment { 
  val species: String,
  val onWarmBlooded: OnWarmBlooded?,
  val onPet: OnPet?
}
data class WarmBloodedAnimalFragment(
  val species: String, 
  val onWarmBlooded: OnWarmBlooded,
  val onPet:OnPet?
): AnimalFragment
data class PetAnimalFragment(
  val species: String,
  val onWarmBlooded: OnWarmBlooded?,
  val onPet: OnPet,
): AnimalFragment
data class WarmBloodedPetAnimalFragment(
  val species: String,
  val onWarmBlooded: OnWarmBlooded,
  val onPet: OnPet,
): AnimalFragment
data class OtherAnimalFragment(
  val species: String,
  val onWarmBlooded: OnWarmBlooded?,
  val onPet: OnPet?,
): AnimalFragment
```
